src/org/labkey/targetedms/chart/ChromatogramDataset.java

```diff
@@ -826,18 +826,36 @@ protected void buildJFreedataset(Chromatogram chromatogram, RtRange chromatogram
             for(int chromatogramIndex = 0; chromatogramIndex < transitionCount; chromatogramIndex++)
             {
                 List<TransitionChromInfo> tChromInfoList = TransitionManager.getTransitionChromInfoList(_pChromInfo.getId(), chromatogramIndex);
-                if(tChromInfoList.isEmpty())
-                    continue;
-                for(TransitionChromInfo tChromInfo: tChromInfoList)
+
+                for (TransitionChromInfo tChromInfo : tChromInfoList)
                 {
                     Transition transition = TransitionManager.get(tChromInfo.getTransitionId(), _user, _container);
 
-                    if(include(transition))
+                    if (include(transition))
                     {
                         tciList.add(new TransChromInfoPlusTransition(tChromInfo, transition));
                     }
                 }
             }
+
+            if (tciList.isEmpty() && _pChromInfo.getTransitionChromatogramIndicesList() != null)
+            {
+                List<Transition> transitions = TransitionManager.getTransitionsForPrecursor(_precursor.getId(), _user, _container);
+                if (transitions.size() != _pChromInfo.getTransitionChromatogramIndicesList().size())
+                {
+                    throw new IllegalStateException("Mismatch in transitions and indices lengths: " + transitions.size() + " vs " + _pChromInfo.getTransitionChromatogramIndicesList().size());
+                }
+                int index = 0;
+                for (Transition transition : transitions)
+                {
+                    if (include(transition))
+                    {
+                        tciList.add(new TransChromInfoPlusTransition(_pChromInfo.makeDummyTransitionChromInfo(index), transition));
+                    }
+                    index++;
+                }
+            }
```

Comment:
looks like a method can be abstracted for the same lines 1125-1140

Reply:
It's very similar code but it's operating on Transitions vs MoleculeTransitions via TransitionManager vs MoleculeTransitionManager, and creating TransChromInfoPlusTransition vs MoleculeTransChromInfoPlusTransition. I just tried coding up a shared version that takes Supplier and BiFunction parameters and it works but it's pretty incomprehensible code. I propose that we leave it as-is.

```
Index: src/org/labkey/targetedms/chart/ChromatogramDataset.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/org/labkey/targetedms/chart/ChromatogramDataset.java	(revision f296b20ed6b2a3fc8c9c35f341828e50929fe1c8)
+++ src/org/labkey/targetedms/chart/ChromatogramDataset.java	(date 1605666252744)
@@ -53,6 +53,8 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
 
 /**
  * User: vsharma
@@ -778,6 +780,30 @@
             return _precursor == null ? 0 : _precursor.getGeneralMoleculeId();
         }
 
+        /** If we didn't find any TransitionChromInfos, see if we can fetch the transitions' indices from the PrecursorChromInfo */
+        protected <TransitionType extends GeneralTransition, TransitionPlusType extends TransitionChromInfoPlusGeneralTransition<TransitionType>>
+        void addDummyTransitionChromInfos(List<TransitionPlusType> tciList, Supplier<List<TransitionType>> transitionGetter,
+                                          BiFunction<Integer, TransitionType, TransitionPlusType> creator)
+        {
+            if (tciList.isEmpty() && _pChromInfo.getTransitionChromatogramIndicesList() != null)
+            {
+                List<TransitionType> transitions = transitionGetter.get();
+                if (transitions.size() != _pChromInfo.getTransitionChromatogramIndicesList().size())
+                {
+                    throw new IllegalStateException("Mismatch in transitions and indices lengths: " + transitions.size() + " vs " + _pChromInfo.getTransitionChromatogramIndicesList().size());
+                }
+                int index = 0;
+                for (TransitionType transition : transitions)
+                {
+                    if (include(transition))
+                    {
+                        tciList.add(creator.apply(index, transition));
+                        index++;
+                    }
+                }
+            }
+        }
+
         @Override
         public void build()
         {
@@ -838,23 +864,9 @@
                 }
             }
 
-            if (tciList.isEmpty() && _pChromInfo.getTransitionChromatogramIndicesList() != null)
-            {
-                List<Transition> transitions = TransitionManager.getTransitionsForPrecursor(_precursor.getId(), _user, _container);
-                if (transitions.size() != _pChromInfo.getTransitionChromatogramIndicesList().size())
-                {
-                    throw new IllegalStateException("Mismatch in transitions and indices lengths: " + transitions.size() + " vs " + _pChromInfo.getTransitionChromatogramIndicesList().size());
-                }
-                int index = 0;
-                for (Transition transition : transitions)
-                {
-                    if (include(transition))
-                    {
-                        tciList.add(new TransChromInfoPlusTransition(_pChromInfo.makeDummyTransitionChromInfo(index), transition));
-                    }
-                    index++;
-                }
-            }
+            Supplier<List<Transition>> transitionGetter = () -> TransitionManager.getTransitionsForPrecursor(_precursor.getId(), _user, _container);;
+            BiFunction<Integer, Transition, TransChromInfoPlusTransition> creator = (i, t) -> new TransChromInfoPlusTransition(_pChromInfo.makeDummyTransitionChromInfo(i), t);
+            addDummyTransitionChromInfos(tciList, transitionGetter, creator);
 
             // Sort according to the ion order used in Skyline
             tciList.sort(new TransChromInfoPlusTransitionComparator());
@@ -1122,22 +1134,9 @@
                 }
             }
 
-            if (tciList.isEmpty() && _pChromInfo.getTransitionChromatogramIndicesList() != null)
-            {
-                List<MoleculeTransition> transitions = MoleculeTransitionManager.getTransitionsForPrecursor(_molPrecursor.getId(), _user, _container);
-                if (transitions.size() != _pChromInfo.getTransitionChromatogramIndicesList().size())
-                {
-                    throw new IllegalStateException("Mismatch in transitions and indices lengths: " + transitions.size() + " vs " + _pChromInfo.getTransitionChromatogramIndicesList().size());
-                }
-                int index = 0;
-                for (MoleculeTransition transition : transitions)
-                {
-                    if (include(transition))
-                    {
-                        tciList.add(new MoleculeTransChromInfoPlusTransition(_pChromInfo.makeDummyTransitionChromInfo(index++), transition));
-                    }
-                }
-            }
+            Supplier<List<MoleculeTransition>> transitionGetter = () -> MoleculeTransitionManager.getTransitionsForPrecursor(_molPrecursor.getId(), _user, _container);
+            BiFunction<Integer, MoleculeTransition, MoleculeTransChromInfoPlusTransition> creator = (i, t) -> new MoleculeTransChromInfoPlusTransition(_pChromInfo.makeDummyTransitionChromInfo(i), t);
+            addDummyTransitionChromInfos(tciList, transitionGetter, creator);
 
             tciList.sort(new MoleculeTransChromInfoPlusTransitionComparator());
 
```